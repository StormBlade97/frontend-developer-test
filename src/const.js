
  export const FETCH_STATE = {
    INITIAL: 'initital',
    PENDING: 'pending',
    SUCCESSFUL: 'successful',
    ERROR: 'error'
  }