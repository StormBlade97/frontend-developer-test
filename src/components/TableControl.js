import React from 'react';
import { FETCH_STATE } from '../const';
import Box from '@material-ui/core/Box';
import Button from '@material-ui/core/Button';
import Typography from '@material-ui/core/Typography';
import CircularProgress from '@material-ui/core/CircularProgress';


const TableControl = ({ fetchState, fetchData }) => {
    const isError = fetchState === FETCH_STATE.ERROR;
    const isLoading = fetchState === FETCH_STATE.PENDING;
    const buttonLabel = isError ? "Retry" : "Load more";

    return <Box p={1}>
        {isError &&
            <Box m={1}>
                <Typography color="error">We had problems fetching your data. Please try again</Typography>
            </Box>
        }
        {
            isLoading ? <CircularProgress ></CircularProgress> :
                <Button variant="contained" color="primary" onClick={fetchData}>
                    {buttonLabel}
                </Button>
        }
    </Box>
}

export default TableControl;