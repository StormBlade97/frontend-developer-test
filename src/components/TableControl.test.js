import React from 'react';
import { shallow } from 'enzyme';
import TableControl from './TableControl';
import { FETCH_STATE } from '../const';
import Button from '@material-ui/core/Button';
import CircularProgress from '@material-ui/core/CircularProgress';
import { Typography } from '@material-ui/core';


describe("TableControl", () => {
    describe("renders", () => {
        describe("when fetching", () => {
            it("should render", () => {
                const wrapper = shallow(<TableControl
                    fetchState={FETCH_STATE.PENDING}
                    fetchData={jest.fn()}
                />)
                expect(wrapper.getElement()).toMatchSnapshot()
            })
            it("should render spinner instead of Button", () => {
                const wrapper = shallow(<TableControl
                    fetchState={FETCH_STATE.PENDING}
                    fetchData={jest.fn()}
                />)
                expect(wrapper.exists(CircularProgress)).toBe(true);
                expect(wrapper.exists(Button)).toBe(false);

            })
        })
        describe("when fetch error", () => {
            it("should render", () => {
                const wrapper = shallow(<TableControl
                    fetchState={FETCH_STATE.ERROR}
                    fetchData={jest.fn()}
                />);
                expect(wrapper.getElement()).toMatchSnapshot();
            });
            it("should render error message and Button", () => {
                const wrapper = shallow(<TableControl
                    fetchState={FETCH_STATE.ERROR}
                    fetchData={jest.fn()}
                />)
                expect(wrapper.exists(Typography)).toBe(true);
                expect(wrapper.exists(Button)).toBe(true);
            })
        })
        describe("when under other fetch state", () => {
            it("should render only button", () => {
                const wrapper = shallow(<TableControl
                    fetchState={FETCH_STATE.SUCCESSFUL}
                    fetchData={jest.fn()}
                />)
                expect(wrapper.getElement()).toMatchSnapshot()
            })
        })
    })
})