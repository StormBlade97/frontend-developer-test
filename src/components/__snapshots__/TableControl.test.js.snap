// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableControl renders when fetch error should render 1`] = `
<ForwardRef(StyledComponent)
  p={1}
>
  <ForwardRef(StyledComponent)
    m={1}
  >
    <ForwardRef(WithStyles)
      color="error"
    >
      We had problems fetching your data. Please try again
    </ForwardRef(WithStyles)>
  </ForwardRef(StyledComponent)>
  <ForwardRef(WithStyles)
    color="primary"
    onClick={[MockFunction]}
    variant="contained"
  >
    Retry
  </ForwardRef(WithStyles)>
</ForwardRef(StyledComponent)>
`;

exports[`TableControl renders when fetching should render 1`] = `
<ForwardRef(StyledComponent)
  p={1}
>
  <ForwardRef(WithStyles) />
</ForwardRef(StyledComponent)>
`;

exports[`TableControl renders when under other fetch state should render only button 1`] = `
<ForwardRef(StyledComponent)
  p={1}
>
  <ForwardRef(WithStyles)
    color="primary"
    onClick={[MockFunction]}
    variant="contained"
  >
    Load more
  </ForwardRef(WithStyles)>
</ForwardRef(StyledComponent)>
`;
